src/components/Table/TableHead/TableHead.jsx

```diff
@@ -0,0 +1,164 @@
+
+const propTypes = {
+  /** Important table options that the head needs to know about */
+  options: PropTypes.shape({ hasRowExpansion: PropTypes.bool, hasRowSelection: PropTypes.bool }),
+  /** List of columns */
+  columns: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.string.isRequired,
+      name: PropTypes.string.isRequired,
+      size: PropTypes.number.isRequired,
+      isSortable: PropTypes.bool,
+      filter: PropTypes.shape({
+        placeholderText: PropTypes.string,
+        options: PropTypes.arrayOf(
+          PropTypes.shape({
+            id: PropTypes.string.isRequired,
+            text: PropTypes.string.isRequired,
+          })
+        ),
+      }),
+    })
+  ).isRequired,
+
+  /** Current state of the table */
+  state: PropTypes.shape({
+    /** Which toolbar is currently active */
+    activeBar: PropTypes.oneOf(['column', 'filter']),
+    /** What's currently selected in the table? */
+    selection: PropTypes.shape({
+      isSelectAllIndeterminate: PropTypes.bool,
+      isSelectAllSelected: PropTypes.bool,
+    }),
+    /** What sorting is currently applied */
+    sort: PropTypes.shape({ direction: PropTypes.string, column: PropTypes.string }),
+    /** What column ordering is currently applied to the table */
+    ordering: PropTypes.arrayOf(
+      PropTypes.shape({
+        columnId: PropTypes.string.isRequired,
+        /* Visibility of column in table, defaults to false */
+        isHidden: PropTypes.bool,
+      })
+    ),
+    /** Optional list of applied column filters */
+    filters: PropTypes.arrayOf(
+      PropTypes.shape({
+        columnId: PropTypes.string.isRequired,
+        value: PropTypes.string.isRequired,
+      })
+    ),
+  }).isRequired,
+  actions: PropTypes.shape({
+    onSelectAll: PropTypes.func,
+    onChangeSort: PropTypes.func,
+    onChangeOrdering: PropTypes.func,
+  }).isRequired,
+};
+
+const defaultProps = {
+  options: {},
+};
+
+const TableHead = ({
+  options,
+  options: { hasRowExpansion, hasRowSelection },
+  columns,
+  state: {
+    selection: { isSelectAllIndeterminate, isSelectAllSelected },
+    sort,
+    activeBar,
+    ordering,
+    filters,
+  },
+  actions: { onSelectAll, onChangeSort, onApplyFilter, onChangeOrdering },
+}) => {
+  const filterBarActive = activeBar === 'filter';
+  const filterBarActiveStyle = { paddingTop: 16 };
+  return (
+    <CarbonTableHead>
+      <TableRow>
+        {hasRowExpansion ? <TableExpandHeader /> : null}
+        {hasRowSelection ? (
+          <TableHeader
+            style={Object.assign(
+              { paddingBottom: '0.5rem' },
+              filterBarActive ? filterBarActiveStyle : {}
+            )}>
+            {/* TODO: Replace checkbox with TableSelectAll component when onChange bug is fixed
+                    https://github.com/IBM/carbon-components-react/issues/1088 */}
+            <Checkbox
+              id="select-all"
```

Comment:
we should pass through the Table id to the child components (like this one) so that all DOM ids are uniquely scoped to the table instance

Reply:
hoping to use css class selectors off the parent table id instead to limit the amount of ids we need to track and set (and reduce the amount of inbound props)

Reply:
alright, that works for me -- as long as everything rendered can be clearly referenced 👍

Reply:
just for reference can you provide the selector I could use to grab the `input` for a column filter in a table?